### DISCUSSION TITLE
ESM dev does not provide an export  named Ref

### DISCUSSION THREAD

**Discussion by productdevbook** [mikro-orm/mikro-orm#4001] [Q&A] [answered]
**Describe the bug**
Esm dev problem.

![image](https://user-images.githubusercontent.com/38668796/216368776-b863edbd-9b14-478b-a0a3-d98171f74d82.png)

```
import { Entity, ManyToOne, Property, Ref } from '@mikro-orm/core'
import { CustomBaseEntity } from './base'
import { DebtEntity } from './debt.entity'

@Entity({ tableName: 'debtLog', schema: 'public' })
export class DebtLogEntity extends CustomBaseEntity {
  @Property({ type: 'number', columnType: 'decimal', default: 0, nullable: false })
  amount!: number

  @Property({ type: 'number', columnType: 'decimal', default: 0, nullable: false })
  beforeLogAmount!: number

 
  @Property({ type: 'number', columnType: 'decimal', default: 0, nullable: false })
  postLogAmount!: number

 
  @Property({ type: 'string', columnType: 'varchar', length: 255, nullable: true })
  note?: string

  @Property({ type: 'boolean', columnType: 'bool', default: false, nullable: false })
  isSystemAdded!: boolean

  @ManyToOne(() => DebtEntity, { eager: false, nullable: false, fieldName: 'debtId', ref: true })
  debt!: Ref<DebtEntity>

  constructor(attrs: Partial<DebtLogEntity> = {}) {
    super()
    Object.assign(this, attrs)
  }
}

```
**Versions**

| Dependency | Version |
| - | - |
| node | v18.13.0 |
| typescript | 4.9.5 |
| mikro-orm | 5.6.8 |
| @mikro-orm/postgresql| 5.6.8 |

**Comment by B4nan** [comment, marked as answer]
This is about your app setup, not about the code. `Ref` is a type, you can't import it in runtime code, as it does not exist on runtime. Maybe try a type import?

```ts
import type { Ref } from '@mikro-orm/core';
```

Btw please don't spam unrelated issues. If you wonder why you did not get any reply to this, it is because you are not providing any details, you literally don't provide any description, you just dump a screenshot somewhere and expect others to do the rest...

**Reply by productdevbook** [reply]
`Btw please don't spam unrelated issues. If you wonder why you did not get any reply to this, it is because you are not providing any details, you literally don't provide any description, you just dump a screenshot somewhere and expect others to do the rest...` 

I find this sentence a bit odd. I deleted my github comment(I deleted it because the place where I wrote github was wrong), wrote the slack side, but I thought it was a bug. You may or may not answer this. I support this project voluntarily and allow many companies to move here and this project is pinned in my profile.

I also write a lot of open source libraries. I've been examining the codes for 1 day, I wonder if Ref is not exported (I appreciate your project and I looked through the codes to even post a pr.) , then I couldn't figure it out and I wrote it. I didn't know what to write in the statement. Because I thought the error was an export problem and I wanted to give details.

I took a screenshot because the error was so obvious to me there. I thought that the name was not exported and it would be solved with the new update by exporting from the structure.

Thanks again for the back and I'll play around with it some more.

**Reply by productdevbook** [reply]
`import type { Ref } from '@mikro-orm/core'`

Thank you very much this is corrected.

**Reply by B4nan** [reply]
This is not about what to write into the description, this is about providing some context. You say it's ESM, but you don't use extensions in imports, so I guess you don't use `tsc` to compile? Or is it some nonstandard tsconfig? Since you say you spent a day looking into the issue, I am sure you have at least some findings to share.

That is the context I am missing here, you can't debug a screenshot. 

Btw I tried my ESM example project and it does not suffer from this. I use `tsc` in there.

**Reply by B4nan** [reply]
I think this is connected to the `isolatedModules` option in TS, do you have that enabled? I believe you should, as it will warn you about places like this. 

https://www.typescriptlang.org/tsconfig#isolatedModules

**Reply by productdevbook** [reply]
no my config

```ts
   "module": "esnext",
    "target": "esnext",
    "lib": [
      "esnext",
      "dom"
    ],
    "baseUrl": ".",
    "outDir": "dist",
    "rootDir": "packages",
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "importHelpers": true,
    "allowJs": true,
    "skipLibCheck": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "sourceMap": true,
    "declaration": true,
    "resolveJsonModule": true,
    "moduleResolution": "node",
```
